# Saloni Patel (A4 - 4.7 - Deriving New Variables)

# Importing Libraries and Data

### Importing Libraries and defining the path

In [3]:
# importing libraries

import pandas as pd
import numpy as np
import os

In [4]:
# Creating Path for ease of code

path = r'C:\Users\psalo\OneDrive\Study\CF - DA Immersion\A4 - Python Fundamentals for Data Analyst\A4 - Instacart Basket Analysis 07-2025'


# Lesson Practice

### 1 - Importing df_ords_prods_merge from Prepared Data Folder

In [5]:
# Importing df_ords_prods_merge from Prepared Data Folder

df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))
print(f"Number of rows imported: {df_ords_prods_merge.shape[0]}")

Number of rows imported: 32404859


### 2 - Checking the imported data shape

In [6]:
# Checking the df_ords_prods_merge shape

df_ords_prods_merge.shape

(32404859, 15)

### 3 - Checking the imported data types

In [7]:
# Checking the df_ords_prods_merge Types

df_ords_prods_merge.dtypes

product_id                  int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_last_order     float64
add_to_cart_order           int64
reordered                   int64
_merge                   category
dtype: object

### 4 - Checking the imported data head

In [10]:
# Checking the df_ords_prods_merge Head

df_ords_prods_merge.head(5)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  \
0   3139998      138    prior            28                   6   
1   1977647      138    prior            30                   6   
2    389851      709    prior             2                   0   
3    652770      764    prior             1                   3   
4   1813452      764    prior             3                   4   

   order_hour_of_day  days_since_last_order  add_to_cart_order  reordered  \
0                 11                    3.0                  5          0   
1                 17                   20.0                  1          1   
2                 21                    6.0                 20          0   
3                 13                    7.0                 10          0   
4                 17                    9.0                 11          1   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

### 5 - Creating a subset for the user defined function

In [11]:
# creating a subset of first million rows 

df = df_ords_prods_merge[:1000000]

In [12]:
# checking the subset shape

df.shape

(1000000, 15)

### 6 - Creating a new user defined variable 'price_label' using if-elif 

In [13]:
# Creating a new user defined variable 'price_label' 

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

### 7 - Creating a new column called price_range. Values will be filled in using the variable 'price_label'

In [14]:
# Creating a new column called price_range. Values will be filled in using the variable 'price_label' - in df

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\psalo\AppData\Local\Temp\ipykernel_11292\2557370238.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [15]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    652638
Low-range product    338018
High range             9344
Name: count, dtype: int64

In [18]:
# looking for the most expensive price - in df

df['prices'].max()


24.5

### 8 - loc function instead of it-elif - on df - another way to avoid doing the user defined function

In [19]:
# testing loc function instad of if-elif statement - on df

df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

C:\Users\psalo\AppData\Local\Temp\ipykernel_11292\2631021952.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [20]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     652638
Low-range product     338018
High-range product      9344
Name: count, dtype: int64

#### NOTE >> output from both user defined function of 'price_range' and loc fuction of 'price_range_loc' are the same

### 9 - loc function on the entire df_ords_prods_merge (not on df that has 1 million rows only)

In [21]:
# product with price that is more than $15 will be 'High-range product'

df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [22]:
# product with price that is more than $5 and less or equal to $15 will be 'Mid-range product'

df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [23]:
# product with price that is less than $5 will be 'Low-range product'

df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [25]:
# checking the values to ensure the functionality

df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

### 10 - TTK - If statement with for-loop

In [33]:
# TTK by printing a repetitive statement

for x in range (2,6):
    print("I am %d years old" % (x))

I am 2 years old
I am 3 years old
I am 4 years old
I am 5 years old


### 11 - how busy each day of the week is - frequency count for the orders_day_of_week

In [36]:
# checking the frequency to understand which day see the highest amount of orders and which one has the lowest

df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

#### NOTE >> 'orders_day_of_week' value 0 is for Saturday and value 4 is for Wednesday. 

In [38]:
# Creating a function to assign 'Busiest day', 'Least busy' and 'Regularly busy' labels for each order following the corresponding value of 'orders_day_of_week'

result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

#### NOTE >> I ran the print result command and the list was too long and so distorting for the notebook view. I have now removed that.

In [43]:
# creating a column in df_ords_prods_merge for this result list so it can be associated with the data frame

df_ords_prods_merge['busiest_day'] = result

In [44]:
# lets check the value count

df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# A4 - 4.7 - Assignment

### 2 - Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

#### as per #11 above we can see that saturday (0) and sunday 91) are top two busiest days and Tuesday (3) and Wednesday (4) are two least busy days. so if foreloop statement will be added to create those corresponding values.

In [55]:
# creating a new list with requested labels for day groups

result1 = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value in (0, 1):
    result1.append("Busiest days")
  elif value in (3, 4):
    result1.append("Least busy days")
  else:
    result1.append("Regularly busy days")

In [56]:
result1

['Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Least busy days',
 'Least busy days',
 'Busiest days',
 'Regularly busy days',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Regularly busy days',
 'Least busy days',
 'Least busy days',
 'Regularly busy days',
 'Least busy days',
 'Regularly busy days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Busiest days',
 'Regularly busy days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Regularly busy days',
 'Busiest days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Lea

In [57]:
# creating a column 'Busiest_days' in df_ords_prods_merge for this result list so it can be associated with the data frame

df_ords_prods_merge['busiest_days'] = result1

### 3 - Check the values of this new column for accuracy. Note any observations in markdown format.


In [58]:
# Checking the values of the new column for accuracy

df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy days    12916111
Busiest days           11864412
Least busy days         7624336
Name: count, dtype: int64

#### NOTE >> The above counts tie back to the breakdown in #11

### 4 - When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [60]:
#Checking the frequecny of the values in variable order_day_of_week

df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

#### Note >> 

1. 9am to 4:59pm = most orders
2. 7am to 8:59am ; 5pm to 10:59pm = Average orders
3. 11pm to 6:59am = Fewest orders

   This information is useful to create the next statement and assign them into a particular group.

In [62]:
# Creating a function to assign 'Most orders', 'Average orders' and 'Fewest orders' labels for each order following the corresponding value of 'orders_hour_of_day'

result2 = []

for value in df_ords_prods_merge["order_hour_of_day"]:
    if value in (9, 10, 11, 12, 13, 14, 15, 16):
        result2.append("Most orders")
    elif value in (7, 8, 17, 18, 19, 20, 21, 22):
        result2.append("Average orders")
    else:
        result2.append("Fewest orders")

In [63]:
# Creating the column 'busiest_period_of_day' and values stored in 'result2' will be assigned to this.

df_ords_prods_merge['busiest_period_of_day'] = result2

### 5 - Print the frequency for this new column.

In [64]:
# Checking the value frequency for the newly created column

df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

#### Note >> The total of the above vaules is 32,404,859 which ties back with other checks.

# Final check and Export of the updated version of df_ords_prods_merge

### Final check

In [65]:
# Checking for the newly added columns in this assignment

df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  \
0   3139998      138    prior            28                   6   
1   1977647      138    prior            30                   6   
2    389851      709    prior             2                   0   
3    652770      764    prior             1                   3   
4   1813452      764    prior             3                   4   

   order_hour_of_day  days_since_last_order  add_to_cart_order  reordered  \
0                 11                    3.0                  5          0   
1                 17                   20.0                  1          1   
2                 21                    6.0                 20          0   
3                 13                    7.0                 10          0   
4                 17                    9.0                 11          1   

  _merge    price_range_loc     busiest_day         busiest_days  \
0   both  Mid-range product  Regularly busy  Regularly busy days   
1   both  Mid-range product  Regularly busy  Regularly busy days   
2   both  Mid-range product     Busiest day         Busiest days   
3   both  Mid-range product  Regularly busy      Least busy days   
4   both  Mid-range product      Least busy      Least busy days   

  busiest_period_of_day  
0           Most orders  
1        Average orders  
2        Average orders  
3           Most orders  
4        Average orders

### Exporting

In [68]:
# Exporting final df in pkl version and printing the number of rows exported for one final verification

num_rows = len(df_ords_prods_merge)  
print(f"{num_rows} rows exported to pkl file.")

df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_post_4_7.pkl'))

32404859 rows exported to pkl file.


In [19]:
# Confirming the merge result 

df_ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64